src/irmin-pack/IO_intf.ml

```diff
@@ -16,27 +16,85 @@
+
+      For a given [pth], there should be at most one read/write instance. There can be
+      multiple readonly instances, but only one read/write instance is allowed. This is
+      not currently enforced (but could be, using lock files for example).  *)
+
   val name : t -> string
+  (** [name t] is the [path] argument provided to {!v} when [t] was created *)  
+
   val append : t -> string -> unit
+  (** [append t s] appends [s] to the end of the file pointed to by [t]. The data is first
+      placed in a buffer, and a flush to file only happens when the amount of data in the
+      buffer exceeds the [auto_flush_limit], which is currently 1M.
+
+      [append] and [set] are the only operations that write to the file, apart from
+      metadata operations. *)
+
   val set : t -> off:int63 -> string -> unit
+  (** [set t ~off s] writes the string [s] at offset [off+header_size]. *)
+
   val read : t -> off:int63 -> bytes -> int
+  (** [read t ~off buf] tries to read [Bytes.length buf] bytes from [t] at offset [off +
+      header_size]; it returns the number of bytes actually read, which may be less than
+      requested if we are near the end of the file *)
+
   val read_buffer : t -> off:int63 -> buf:bytes -> len:int -> int
+  (** [read_buffer t ~off ~buf ~len] tries to read [len] bytes from [t] at offset
+      [off+header_size] *)
+
   val offset : t -> int63
+  (** [offset t] returns the offset field of [t], which for the read/write instance is the
+      position at which data is written using {!append}. *)
+
   val force_offset : t -> int63
+  (** [force_offset t] returns the last offset for data that was reliably synced to disk
+      from the internal buffer, and as a side effect it updates the [t.offset] field to
+      this value. This should {b only} be called for read-only instances, where it enables
+      the instance to detect that there is more data to read. It does not make sense to
+      call this function for the read-write instance, but this is not currently
+      enforced. *)
+
   val readonly : t -> bool
   val flush : t -> unit
   val close : t -> unit
+
   val exists : string -> bool
+  (** [exists pth] is just [Sys.file_exists pth]; it does not check that the [pth] points
+      to a valid backing file that can be opened with {!v}. FIXME why include in the
+      interface? *)
+
   val size : t -> int
+  (** Returns the real size of the underlying file, including the header data. *)
 
   val truncate : t -> unit
-  (** Sets the length of the underlying IO to be 0, without actually purging the
-      associated data. *)
+  (** [truncate t] sets the internal field [t.offset] to 0 and [t.flushed] to point just
+      after the header data. The underlying file is not truncated - the data is still
+      there. A future flush will persist [t.offset] in the underlying file header.
+
+      [t.flushed] is an internal field that is used to trigger flushing from the internal
+      buffer *)
```

Comment:
I think I preferred the original comment here w.r.t. getting an understanding of the semantics. Perhaps we could extend that one to say that the purging is delayed until the next internal flush? The field names and implementation details seem likely to become outdated and make it difficult to check that the function actually behaves as required by call-sites.

Reply:
I agree that it is not nice to expose the internal implementation details. But it wasn't clear to me what this function is actually supposed to do. It sets some internal fields which has various consequences. But (for example) the raw offset is not updated till some time later. This means that, until a flush occurs, there is garbage data that will be read as real data if a crash occurs. Basically, I am not sure what this function is supposed to do, without a flush of the underlying raw offset.

Reply:
The "bad" scenario is: start with huge file; "truncate"; write some data, but not enough to trigger a flush; crash; restart; now the offset points to the end of the huge file, but we thought we were writing new data at the beginning of the file.

Reply:
I agree that this is a weird function, and we should consider separately whether the callers can actually tolerate the flush being lost due to laziness (I suspect they cannot, and so this bad case is indeed a bug).

w.r.t. to documenting the current behaviour though, it seems fine to me to say that the actual truncation is delayed until the next flush.

Reply:
You mentioned that RO instances use the offset to trigger resyncs of dictionary and index. https://github.com/mirage/irmin/blob/1d83f084815d11140f9b77fcb4732851bb6dec18/src/irmin-pack/pack_store.ml#L556-L558 

The code says something like: `if offset > former_offset then ...`. How does this work with truncate, which presumably sets offset to 0... why not `if offset <> former_offset then...` ?

Reply:
Or, indeed, how does this work with `set`, which mutates the file data but doesn't alter offset at all.